### PR TITLE
[feature](agg) support aggregate function group_array_intersect

### PR DESCRIPTION
**Pull request by superdiaodiao**
## Proposed changes

Issue Number: #31737 

Consider this table named group_array_intersect_test, which has id and c_array_int columns:
```
+------+-----------------+
| id   | c_array_int     |
+------+-----------------+
|    0 | [0]             |
|    6 | [null]          |
|   12 | [12, null, 13]  |
|   14 | [12, 13]        |
+------+-----------------+
```
we can use **group_array_intersect** to get the intersect element(s) from given array like these queries:
```
mysql> select group_array_intersect(c_array_int) from group_array_intersect_test where id in (6, 12);
+------------------------------------+
| group_array_intersect(c_array_int) |
+------------------------------------+
| [null]                             |
+------------------------------------+
1 row in set (0.02 sec)

mysql> select group_array_intersect(c_array_int) from group_array_intersect_test where id in (14, 12);
+------------------------------------+
| group_array_intersect(c_array_int) |
+------------------------------------+
| [13, 12]                           |
+------------------------------------+
1 row in set (0.01 sec)

mysql> select group_array_intersect(c_array_int) from group_array_intersect_test where id in (0, 6);
+------------------------------------+
| group_array_intersect(c_array_int) |
+------------------------------------+
| []                                 |
+------------------------------------+
1 row in set (0.00 sec)
```
Furthermore, this function supports **all kinds of array type**, not only array(int) mentioned above, but also array(varchar()), array(date) and so on.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run feut

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/GroupArrayIntersect.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+
+/**
+ * AggregateFunction 'group_array_intersect'.
+ */
+public class GroupArrayIntersect extends AggregateFunction
+        implements UnaryExpression, ExplicitlyCastableSignature, AlwaysNotNullable {
+
+    public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
+            FunctionSignature.ret(ArrayType.of(new FollowToAnyDataType(0)))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/GroupArrayIntersect.java:42]:
```suggestion
            FunctionSignature.retArgType(0)
```

Comment (superdiaodiao) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/GroupArrayIntersect.java:42]: ok

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+template <typename T>
+struct AggregateFunctionGroupArrayIntersectData {
+    using NullableNumericOrDateSetType = NullableNumericOrDateSet<T>;
+    using Set = std::shared_ptr<NullableNumericOrDateSetType>;
+
+    AggregateFunctionGroupArrayIntersectData()
+            : value(std::make_shared<NullableNumericOrDateSetType>()) {}
+
+    Set value;
+    UInt64 version = 0;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:93]: `bool init = false;`

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+            for (size_t i = 0; i < arr_size; ++i) {
+                const bool is_null_element =
+                        is_column_data_nullable && col_null->is_null_at(offset + i);
+                const T* src_data =
+                        is_null_element ? nullptr : &(nested_column_data->get_element(offset + i));
+
+                set->insert(src_data);
+            }
+        } else if (set->size() != 0 || set->contain_null()) {
+            typename State::Set new_set =
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:169]: `auto new_set = std::make_shared<typename State::NullableNumericOrDateSetType>;`

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+public:
+    NullableNumericOrDateSet() { this->_null_aware = true; }
+
+    void change_contains_null_value(bool target_value) { this->_contains_null = target_value; }
+};
+
+template <typename T>
+struct AggregateFunctionGroupArrayIntersectData {
+    using NullableNumericOrDateSetType = NullableNumericOrDateSet<T>;
+    using Set = std::shared_ptr<NullableNumericOrDateSetType>;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:87]: consider use unique_ptr

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h (hunk 2 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h]: ok

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h (hunk 3 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:169]: ok

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h (hunk 4 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h]: ok

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+class Arena;
+class BufferReadable;
+class BufferWritable;
+} // namespace doris::vectorized
+
+namespace doris::vectorized {
+
+/// Only for changing Numeric type or Date(DateTime)V2 type to PrimitiveType so that to inherit HybridSet
+template <typename T>
+constexpr PrimitiveType TypeToPrimitiveType() {
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:49]: type_to_primitive_type()

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+// and modified by Doris
+
+#include "vec/aggregate_functions/aggregate_function_group_array_intersect.h"
+
+namespace doris::vectorized {
+
+IAggregateFunction* create_with_extra_types(const DataTypePtr& nested_type,
+                                            const DataTypes& argument_types) {
+    WhichDataType which(nested_type);
+    if (which.idx == TypeIndex::Date || which.idx == TypeIndex::DateV2) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.cpp:28]: Maybe we should only support DateV2 and DateTimeV2?

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+        const ColVecType* nested_column_data = nullptr;
+
+        if (is_column_data_nullable) {
+            auto const_col_data = const_cast<IColumn*>(&column_data);
+            col_null = static_cast<ColumnNullable*>(const_col_data);
+            nested_column_data = &assert_cast<const ColVecType&>(col_null->get_nested_column());
+        } else {
+            nested_column_data = &static_cast<const ColVecType&>(column_data);
+        }
+
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:157]: This method is too complicated. Maybe some logic can be move into the member function of AggregateFunctionGroupArrayIntersectData?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 8]
```diff
@@ -1503,7 +1505,9 @@ private void initAggregateBuiltins() {
             addBuiltin(
                     AggregateFunction.createBuiltin(GROUP_ARRAY, Lists.newArrayList(t, Type.INT), new ArrayType(t),
                             t, "", "", "", "", "", true, false, true, true));
-
+            addBuiltin(
+                    AggregateFunction.createBuiltin(GROUP_ARRAY_INTERSECT, Lists.newArrayList(new ArrayType(t)),
```
Comment (BiteTheDDDDt) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1509]: No need to add symbols to traditional planner

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h (hunk 5 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h]: done

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.cpp (hunk 6 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.cpp]: ok

File: be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h (hunk 7 above)
Comment (superdiaodiao) [reply be/src/vec/aggregate_functions/aggregate_function_group_array_intersect.h:183]: done

**Comment by superdiaodiao** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run p0

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]

**Pull request merged**